internal/aws/xray/telemetry/recorder.go

```diff
@@ -0,0 +1,327 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telemetry // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/xray/telemetry"
+
+import (
+	"errors"
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/xray"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/awsutil"
+	awsxray "github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/xray"
+)
+
+const (
+	envAWSHostname   = "AWS_HOSTNAME"
+	envAWSInstanceID = "AWS_INSTANCE_ID"
+
+	queueSize       = 30
+	batchSize       = 10
+	defaultInterval = time.Minute
+)
+
+type Recorder interface {
+	// Start send loop.
+	Start()
+	// Stop send loop.
+	Stop()
+	// RecordSegmentsReceived adds the count to the record for the current period.
+	RecordSegmentsReceived(count int)
+	// RecordSegmentsSent adds the count to the record for the current period.
+	RecordSegmentsSent(count int)
+	// RecordSegmentsSpillover adds the count to the record for the current period.
+	RecordSegmentsSpillover(count int)
+	// RecordSegmentsRejected adds the count to the record for the current period.
+	RecordSegmentsRejected(count int)
+	// RecordConnectionError categorizes the error and increments the count by one for the current period.
+	RecordConnectionError(err error)
+}
+
+type telemetryRecorder struct {
+	resourceARN string
+	instanceID  string
+	hostname    string
+
+	// client is used to send the records.
+	client awsxray.XRayClient
+	// record is the pointer to the count metrics for the current period.
+	record *xray.TelemetryRecord
+	// mu is the lock used when updating the record. Primarily exists to
+	// prevent writes while swapping the record out in cutoff.
+	mu sync.Mutex
+
+	// queue is used to keep records that failed to send for retry during
+	// the next period.
+	queue chan *xray.TelemetryRecord
```

Comment:
A channel to queue failed events feels like a potential hazard, in the event that you're heavily throttled, there is a risk that your retry queue is going to fill and due it being fixed size, could mean actual dropping of data.

Would an atomic slice be better here instead?

Reply:
I would prefer that it drop older data than grow unbounded. I did consider using an atomic slice earlier.